nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -101,9 +101,43 @@ declare_saga_actions! {
     FIND_NEW_REGION -> "new_dataset_and_region" {
         + rsrss_find_new_region
     }
+    // One of the common sharp edges of sagas is that the compensating action of
+    // a node does _not_ run if the forward action fails. Said another way, for
+    // this node:
+    //
+    // EXAMPLE -> "output" {
+    //   + forward_action
+    //   - forward_action_undo
+    // }
+    //
+    // If `forward_action` fails, `forward_action_undo` is never executed.
+    // Forward actions are therefore required to be atomic, in that they either
+    // fully apply or don't apply at all.
+    //
+    // Sagas with nodes that ensure multiple regions exist cannot be atomic
+    // because they can partially fail (for example: what if only 2 out of 3
+    // ensures succeed?). In order for the compensating action to be run, it
+    // must exist as a separate node that has a no-op forward action:
+    //
+    // EXAMPLE_UNDO -> "not_used" {
+    //   + noop
+    //   - forward_action_undo
+    // }
+    // EXAMPLE -> "output" {
+    //   + forward_action
+    // }
+    //
+    // This saga will only ever ensure that a single region exists, so you might
+    // think you could get away with a single node that combines the forward and
+    // compensating action - you'd be mistaken! The Crucible agent's region
+    // ensure is not atomic in all cases: if the region fails to create, it
+    // enters the `failed` state, but is not deleted. Nexus must clean these up.
+    NEW_REGION_ENSURE_UNDO -> "not_used" {
+        + rsrss_noop
+        - rsrss_new_region_ensure_undo
```

Comment:
I think this approach works, so this is more of a curiosity than anything, but: would it make sense to model this as one node that kicks off region creation and one that waits for region creation to resolve? (That is, can we split `ensure_region_in_dataset` into "issue operation" and "wait for result" pieces and do one in each node?) Then the undo action for the first node can still call `ensure_undo`, and (I'd imagine) the second doesn't need an undo action at all.

Reply:
If we split like this, you're right that the second one wouldn't need an undo action, but I think the first one would still need to follow this pattern of having a previous node be responsible for the undo action, otherwise we'd end up with this same bug?

Reply:
Depends on how the first action works, I think: it would only be able to fail if it also knew there was no work to undo (i.e. it'd need to get back some sort of error status from its callee that implies that the region is not created and will not be created).